src/util/psbt/mod.rs

```diff
@@ -209,6 +218,473 @@ impl PartiallySignedTransaction {
+            };
+
+            let pk = sk.public_key(secp);
+            sigs.push((pk, sig));
+        }
+
+        for (pk, sig) in sigs.iter() {
+            self.inputs[input_index]
+                .partial_sigs
+                .insert(*pk, *sig);
+        }
+
+        Ok(())
+    }
+
+    /// Returns the sighash message to sign an ECDSA input along with the sighash type.
+    ///
+    /// Uses the [`EcdsaSighashType`] from this input if one is specified. If no sighash type is
+    /// specified uses [`EcdsaSighashType::All`]. This function does not support scripts that
+    /// contain `OP_CODESEPARATOR`.
+    ///
+    /// # Panics
+    /// If called with a taproot input.
+    pub fn sighash_ecdsa<T: Deref<Target=Transaction>>(
+        &self,
+        input_index: usize,
+        cache: &mut SighashCache<T>,
+    ) -> Result<(Message, EcdsaSighashType), SignError> {
+        assert_eq!(self.signing_algorithm(input_index), Ok(SigningAlgorithm::Ecdsa));
+
+        let input = self.checked_input(input_index)?;
+        let utxo = self.spend_utxo(input_index)?;
+        let spk = &utxo.script_pubkey; // scriptPubkey for input spend utxo.
+
+        let hash_ty = input.ecdsa_hash_ty()
+            .map_err(|_| SignError::InvalidSighashType)?; // Only support standard sighash types.
+
+        let output_type = self.output_type(input_index)?;
+
+        let is_wpkh = output_type == OutputType::Wpkh;
+        let is_nested_wpkh = output_type == OutputType::ShWpkh;
+
+        let sighash = if output_type.is_segwitv0() {
+            if is_wpkh || is_nested_wpkh {
+                let script_code = if is_wpkh {
+                    Script::p2wpkh_script_code(&spk).ok_or(SignError::NotWpkh)?
+                } else {
+                    Script::p2wpkh_script_code(input.redeem_script.as_ref().expect("checked above"))
+                        .ok_or(SignError::NotWpkh)?
+                };
+                cache.segwit_signature_hash(input_index, &script_code, utxo.value, hash_ty)?
+            } else {
+                let script_code = input.witness_script.as_ref().ok_or(SignError::MissingWitnessScript)?;
+                cache.segwit_signature_hash(input_index, script_code, utxo.value, hash_ty)?
+            }
+        } else {
+            let script_code = if spk.is_p2sh() {
+                input.redeem_script.as_ref().ok_or(SignError::MissingRedeemScript)?
+            } else {
+                &spk
+            };
+            cache.legacy_signature_hash(input_index, script_code, hash_ty.to_u32())?
+        };
+
+        Ok((Message::from_slice(&sighash).expect("sighashes are 32 bytes"), hash_ty.into()))
+    }
+
+    /// Returns the spending utxo for this PSBT's input at `input_index`.
+    pub fn spend_utxo(&self, input_index: usize) -> Result<&TxOut, SignError> {
+        let input = self.checked_input(input_index)?;
+        let utxo = if let Some(witness_utxo) = &input.witness_utxo {
+            &witness_utxo
```

Comment:
We can remove the redundant borrow here so it's just `witness_utxo`

Reply:
Face palm, `cargo clippy` is a thing.